README.md
jsmstrade
=========

A minimal GUI tool to send SMS via smstrade.de


CHANGELOG
Mon Feb 8 - Changes from v0.2 -> v0.3
	Fixed a major bug. JSMSTrade could not deliver any SMS messages.

Mon Feb 2 - Changes from v0.1 -> v0.2
	Default language is now english (used to be german)
	Bugfix: If a message > 160 chars has been typed, the form will not
	freeze any more


scripts/checklinelengths.sh
#!/bin/sh
# 
# Copyright (c) 2008 Paul C. Buetow, vs@dev.buetow.org
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

MAXLEN=80

for i in `find ./sources -name \*.java`
do
	awk -v MAXLEN=$MAXLEN -v file=$i '{
		if (length > MAXLEN)
			print file " line " NR " is " length " chars long "
	}' $i
done


sources/smstrade/SFrame.java
/* JSMSTrade v0.3
 * Copyright (c) 2008, 2009 Dipl.-Inform. (FH) Paul C. Buetow
 * jsmstrade@dev.buetow.org - http://jsmstrade.buetow.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * All icons of the icons/ folder are 	under a Creative Commons
 * Attribution-Noncommercial-Share Alike License a CC-by-nc-sa.
 */

package smstrade;

import java.awt.*;
import java.awt.event.*;
import javax.swing.JFrame;

/**
 * The class SFrame. All frames of the tool extend this SFrame class.
 * This class makes sure that all 'subwindows' get closed if its parent gets
 * closed. And it also makes sure to open new windows relative to its parent.
 *
 * @author Paul C. Buetow
 */
public class SFrame extends JFrame {
    /** The serial version uid */
    private static final long serialVersionUID = 1L;

    /** The Constant X_LOCATION_OFFSET. */
    private final static int X_LOCATION_OFFSET = 40;

    /** The Constant Y_LOCATION_OFFSET. */
    private final static int Y_LOCATION_OFFSET = 80;

    /** The parent window/component. */
    private Component parent;

    /** True, if the current window will get disposed with its parent.  */
    private boolean dispose;

    /**
     * Instantiates a SFrame object.
     *
     * @param title the title
     * @param parent the parent
     */
    public SFrame(String title, Component parent) {
        super(title);
        init(parent);
    }

    /**
     * Instantiates a new SFrame object.
     *
     * @param title the title
     */
    public SFrame(String title) {
        super(title);
        init(null);
    }

    /**
     * Inits the SFrame.
     *
     * @param parent the parent
     */
    private void init(Component parent) {
        this.parent = parent;
        this.dispose = false;
    }

    /**
     * Dispose with its parent.
     */
    public void disposeWithParent() {
        if (!dispose && parent != null && parent instanceof Window) {
            Window window = (Window) parent;
            window.addWindowListener(new WindowAdapter() {
                public void windowClosed(WindowEvent we) {
                    SFrame.this.dispose();
                }
            });
        }
        dispose = true;
    }

    /**
     * Sets the correct location of the window.
     */
    private void setCorrectLocation() {
        int x = 0, y = 0;
        final Dimension screenSize =
            Toolkit.getDefaultToolkit().getScreenSize();

        if (parent == null) {
            x = (int) (screenSize.width - getWidth()) / 2;
            y = 50;//(int) (screenSize.height - getHeight()) / 2;

        } else {
            final Point location = parent.getLocation();
            x = (int) location.getX() + X_LOCATION_OFFSET;
            y = (int) location.getY() + Y_LOCATION_OFFSET;
        }

        if (x + super.getWidth() >= screenSize.width)
            x = screenSize.width - super.getWidth();
        else if (x < 0)
            x = 0;

        if (y + super.getHeight() >= screenSize.height)
            y = screenSize.height - super.getHeight();

        super.setLocation(x, y);
    }

    /* (non-Javadoc)
     * @see java.awt.Window#setSize(int, int)
     */
    public void setSize(int width, int height) {
        super.setSize(width, height);
        setCorrectLocation();
    }
}


sources/smstrade/SPrefs.java
/* JSMSTrade v0.3
 * Copyright (c) 2008, 2009 Dipl.-Inform. (FH) Paul C. Buetow
 * jsmstrade@dev.buetow.org - http://jsmstrade.buetow.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * All icons of the icons/ folder are 	under a Creative Commons
 * Attribution-Noncommercial-Share Alike License a CC-by-nc-sa.
 */

package smstrade;

import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.util.*;
import javax.swing.*;
import javax.swing.event.*;

/**
 * The class SPrefs.
 *
 * @author Paul C. Buetow
 */
public class SPrefs extends SFrame {
    /** The serial version uid */
    private static final long serialVersionUID = 1L;

    /** The options map */
    private HashMap<String,String> options = null;

    /** The text area */
    private JTextArea textArea = new JTextArea();

    /** The button panel */
    private JPanel buttonPanel = new JPanel();

    /** The ok button */
    private JButton okButton = new JButton("OK");

    /** The save button */
    private JButton saveButton = new JButton("Save");

    /**
     * Instantiates a new SMain object.
     */
    public SPrefs(Component parent, HashMap<String,String> options) {
        super("Preferences", parent);
        this.options = options;

        disposeWithParent();
        setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        setSize(300, 150);
        setResizable(false);

        fillContentPane();
        setVisible(true);
    }

    /**
     * Fills the content pane
     */
    private void fillContentPane() {
        JPanel contentPanel = (JPanel) getContentPane();
        contentPanel.setLayout(new BoxLayout(contentPanel, BoxLayout.Y_AXIS));

        if (options.containsKey("URL"))
            textArea.setText(options.get("URL"));
        else
            textArea.setText(SMain.DEFAULT_URL);

        textArea.setLineWrap(true);

        contentPanel.add(textArea);
        contentPanel.add(createButtonPanel());
    }

    /**
     * Creates the button panel
     *
     * @return The button panel
     */
    private JPanel createButtonPanel() {
        ActionListener listener = new ActionListener() {
            public void actionPerformed(ActionEvent ae) {
                JButton source = (JButton) ae.getSource();
                String text = source.getText();

                if (text.equals("OK")) {
                    save();
                    dispose();

                } else if (text.equals("Save")) {
                    save();
                }
            }

            private void save() {
                options.put("URL", textArea.getText());

                try {
                    FileOutputStream fos =
                        new FileOutputStream(SMain.SAVE_FILE);
                    ObjectOutputStream oos =
                        new ObjectOutputStream(fos);

                    oos.writeObject(options);

                } catch (Exception e) {
                    System.err.println(e);
                }
            }
        };

        buttonPanel.add(okButton);
        buttonPanel.add(saveButton);

        okButton.addActionListener(listener);
        saveButton.addActionListener(listener);

        return buttonPanel;
    }
}


sources/smstrade/SMain.java
/* JSMSTrade v0.3
 * Copyright (c) 2008, 2009 Dipl.-Inform. (FH) Paul C. Buetow
 * jsmstrade@dev.buetow.org - http://jsmstrade.buetow.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * All icons of the icons/ folder are 	under a Creative Commons
 * Attribution-Noncommercial-Share Alike License a CC-by-nc-sa.
 */

package smstrade;

import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.net.*;
import java.util.*;
import javax.swing.*;
import javax.swing.event.*;

/**
 * The class SMain.
 *
 * @author Paul C. Buetow
 */
public class SMain extends SFrame {
    /** The serial version uid */
    private static final long serialVersionUID = 1L;

    /** The max length of a SMS message */
    private static final int MESSAGE_MAX_LENGTH = 160;

    /** The program version */
    private static final double VERSION = 0.3;

    /** The save file */
    public static final String SAVE_FILE = "jsmstrade.dat";

    /** The default URL */
    public static final String DEFAULT_URL =
        "https://gateway.smstrade.de?key=KEY&to=TO&route=basic&message=";

    /** The options map */
    private HashMap<String,String> options = null;

    /** The text area */
    private JTextArea textArea = new JTextArea();

    /** The send button */
    private JButton sendButton = new JButton("Send it");

    /** The clear button */
    private JButton clearButton = new JButton("Delete");

    /** The counter label button */
    private JLabel counterLabel = new JLabel(" 160");

    /** The counter text label button */
    private JLabel counterTextLabel = new JLabel("Chars");

    /** The button panel */
    private JPanel buttonPanel = new JPanel();

    /** The menu bar */
    private JMenuBar menuBar = new JMenuBar();

    /** The file menu */
    private JMenu fileMenu = new JMenu("File");

    /**
     * Instantiates a new SMain object.
     */
    public SMain(HashMap<String,String> options) {
        super("JSMSTrade v" + VERSION);
        this.options = options;

        setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        setSize(300, 150);
        setResizable(false);

        fillMenuBar();
        fillContentPane();
        setVisible(true);
    }

    /**
     * Fills the menu bar
     */
    private void fillMenuBar() {
        menuBar.add(fileMenu);

        ActionListener listener = new ActionListener() {
            public void actionPerformed(ActionEvent ae) {
                JMenuItem source = (JMenuItem) ae.getSource();
                String text = source.getText();

                if (text.equals("Preferences")) {
                    new SPrefs(SMain.this, options);

                } else if (text.equals("Quit")) {
                    System.exit(0);
                }
            }
        };

        JMenuItem prefsItem = new JMenuItem("Preferences");
        prefsItem.addActionListener(listener);
        fileMenu.add(prefsItem);

        fileMenu.addSeparator();
        JMenuItem exitItem = new JMenuItem("Quit");
        exitItem.addActionListener(listener);
        fileMenu.add(exitItem);

        setJMenuBar(menuBar);
    }

    /**
     * Fills the content pane
     */
    private void fillContentPane() {
        JPanel contentPanel = (JPanel) getContentPane();
        contentPanel.setLayout(new BoxLayout(contentPanel, BoxLayout.Y_AXIS));
        contentPanel.add(textArea);
        contentPanel.add(createButtonPanel());

        textArea.addKeyListener(new KeyListener() {
            public void keyTyped(KeyEvent ke) {
                updateCounter(ke);
            }

            public void keyReleased(KeyEvent ke) {
                updateCounter(ke);
            }

            public void keyPressed(KeyEvent ke) {
                updateCounter(ke);
            }

            private void updateCounter(KeyEvent ke) {
                int length = textArea.getText().length();

                if (length > MESSAGE_MAX_LENGTH) {
                    int diff = length - MESSAGE_MAX_LENGTH;
                    textArea.setText(
                        textArea.getText().substring(0, length-diff));
                    ke.consume();

                } else {
                    buttonPanel.remove(2);
                    counterLabel = new JLabel(" "+(160-length));
                    buttonPanel.add(counterLabel, 2);
                    buttonPanel.updateUI();
                    if (length == 0) {
                        clearButton.setEnabled(false);
                        sendButton.setEnabled(false);
                    } else {
                        clearButton.setEnabled(true);
                        sendButton.setEnabled(true);
                    }
                }
            }
        });

        textArea.setLineWrap(true);
        textArea.setWrapStyleWord(true);
    }

    /**
     * Creates the button panel
     *
     * @return The button panel
     */
    private JPanel createButtonPanel() {
        ActionListener listener = new ActionListener() {
            public void actionPerformed(ActionEvent ae) {
                JButton source = (JButton) ae.getSource();
                String text = source.getText();

                if (text.equals("Send it")) {
                    String message = textArea.getText();
                    textArea.setText("");
                    buttonPanel.remove(2);
                    counterLabel = new JLabel(" 160");
                    buttonPanel.add(counterLabel, 2);
                    buttonPanel.updateUI();
                    clearButton.setEnabled(false);
                    sendButton.setEnabled(false);

                    try {
                        String urlString = options.get("URL")
                                           + URLEncoder.encode(
                                               message, "ISO-8859-1");

                        URL url = new URL(urlString);
                        URLConnection conn = url.openConnection();
                        InputStream is = conn.getInputStream();

                        int ch = 0;
                        while ((ch = is.read()) != -1) {
                        }
                        is.close();

                    } catch (Exception e) {
                        System.err.println(e);
                    }
			
                    buttonPanel.updateUI();
                } else if (text.equals("Delete")) {
                    textArea.setText("");
                    buttonPanel.remove(2);
                    counterLabel = new JLabel(" 160");
                    buttonPanel.add(counterLabel, 2);
                    buttonPanel.updateUI();
                    clearButton.setEnabled(false);
                    sendButton.setEnabled(false);
                }
            }
        };

        buttonPanel.add(sendButton);
        buttonPanel.add(clearButton);
        buttonPanel.add(counterLabel);
        buttonPanel.add(counterTextLabel);

        sendButton.addActionListener(listener);
        clearButton.addActionListener(listener);

        clearButton.setEnabled(false);
        sendButton.setEnabled(false);

        return buttonPanel;
    }

    /**
     * The main method.
     *
     * @param args the arguments
     */
    @SuppressWarnings("unchecked")
    public static void main(String[] args) {
        try {
            UIManager.setLookAndFeel(
                UIManager.getCrossPlatformLookAndFeelClassName());
        } catch (Exception e) { }

        File file = new File(SMain.SAVE_FILE);
        HashMap<String,String> options = null;

        if (file.exists()) {
            try {
                FileInputStream fis = new FileInputStream(file);
                ObjectInputStream ois = new ObjectInputStream(fis);
                options = (HashMap<String,String>) ois.readObject();

            } catch (Exception e) {
                System.err.println(e);
            }

        } else {
            options = new HashMap<String, String>();
        }

        new SMain(options);
    }
}
